### PR TITLE
Implementation of the Least Squares Equation Solver 

### PR DESCRIPTION
**Pull request by sofianehaddad**
Least Squares solver for a system of non-linear equations

Continuation of the work done in #2601 with some changes to come

 * C++ classes and tests
 * Python bindings + tests
 * API doc

### CODE REVIEW

File: python/test/t_LeastSquaresEquationsSolver_std.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import openturns as ot
+import openturns.experimental as otexp
+import openturns.testing as ott
+
+ot.TESTPREAMBLE()
+
+# Analytical construction */
+input = ot.Description(2)
+input = ["x", "y"]
+formulas = ot.Description(2)
```
Comment (jschueller) [inline python/test/t_LeastSquaresEquationsSolver_std.py:12]: no need to use ot.Description

File: python/test/t_LeastSquaresEquationsSolver_std.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+formulas = ["y*x-sin(2*x)", "1 + cos(y) + x"]
+analytical = ot.SymbolicFunction(input, formulas)
+algo = otexp.LeastSquaresEquationsSolver()
+algo.setResidualError(1e-10)
+algo.setMaximumCallsNumber(1000)
+starting_point = ot.Point([2.0, 1.0])
```
Comment (jschueller) [inline python/test/t_LeastSquaresEquationsSolver_std.py:18]: nor Point

File: lib/test/t_LeastSquaresEquationsSolver_std.cxx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    algo.setResidualError(1e-5);
+    algo.setMaximumCallsNumber(1000),
+    fullprint << "algo=" << algo << std::endl;
+    Point startingPoint = {2.0, 1.0};
+    const Point optimalValue(2);
+    OSS oss;
+    oss.setPrecision(5);
+    Point solution(algo.solve(analytical, startingPoint));
+    fullprint << "Solve " << formulas << "= [0,0] for " << input <<std::endl;
+    fullprint << "[x,y] = " << String(oss << solution) << std::endl;
```
Comment (jschueller) [inline lib/test/t_LeastSquaresEquationsSolver_std.cxx:49]: String(oss << solution) ? I think the String can be dropped here

File: lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+{
+  OSS oss;
+  oss << "class=" << LeastSquaresEquationsSolver::GetClassName()
+      << " derived from " << SolverImplementation::__repr__();
+  return oss;
+}
+
+void LeastSquaresEquationsSolver::setOptimizationAlgorithm(const OptimizationAlgorithm & algorithm)
+{
+  useDefaultOptimizationAlgorithm_ = false;
```
Comment (jschueller) [inline lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:69]: I dont think this flag is useful: we can just instanciate the default least squares optimizer in the ctor, see QuantileMatchingFactory

Comment (sofianehaddad) [reply lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:69]: ok will have a look, thanks

File: lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  }
+  catch (const InvalidArgumentException &)
+  {
+    LOGWARN("Default optimization algorithm could not solve the least squares problem. Trying to set up a new one...");
+    solver = OptimizationAlgorithm::Build(lsqProblem);
+  }
+  solver.setProblem(lsqProblem);
+  solver.run();
+  callsNumber = solver.getResult().getCallsNumber();
+  callsNumber_ = callsNumber;
```
Comment (jschueller) [inline lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:122]: seems the callsNumber variable is not very useful

Comment (sofianehaddad) [reply lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx]: yes

File: lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  return solver_;
+}
+
+/* Solve attempt to find one root to the system of non-linear equations function(x) = 0
+   given a starting point x with a least square optimization method.
+*/
+Point LeastSquaresEquationsSolver::solve(const Function & function,
+                                         const Point & startingPoint) const
+{ 
+  const Interval bounds;
```
Comment (regislebrun) [inline lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:89]: Shouldn't the bounds include the starting point? And what about the dimension of the default Interval wrt the dimension of the starting point or the input dimension of the function?
IMO, something like
```
const Interval bounds(Point(startingPoint.getDimension(), -SpecFunc::MaxScalar), Point(startingPoint.getDimension(), SpecFunc::MaxScalar));
```
would give more sense to these bounds.

**Review by regislebrun — CHANGES_REQUESTED** [review]
Almost good for me. The default bounds should be changed before to get a green light.

File: lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx (hunk 6 above)
Comment (sofianehaddad) [reply lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:89]: `Interval(0)` is used by the `OptimizationProblem` thus it is in coherency with the rest of the code
Also we check that `Interval.getBounds.getSize` is cimpatible with `objective.getInputFunction`

Comment (jschueller) [reply lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:89]: these are bounds of dimension 0 which means no bounds for the optim solver

Comment (regislebrun) [reply lib/src/Base/Solver/LeastSquaresEquationsSolver.cxx:89]: Correct, so it is ok for me!

**Review by regislebrun — APPROVED** [review]

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 7]
```diff
@@ -0,0 +1,81 @@
+%feature("docstring") OT::LeastSquaresEquationsSolver
+"Solve systems of non-linear equations using a least squares method.
+
```
Comment (jschueller) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:3]: its missing the experimental warning we usually put here, see other experimental classes
also could you add it to the changelog ?

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    and is set to :math:`0`.
+maximumFunctionEvaluation : int
+    The maximum number of evaluations of the function. 
+    Default value is defined in the `ResourceMap` via the `Solver-DefaultMaximumFunctionEvaluation`
+    and is set to :math:`100`.
+
+See also
+--------
+openturns.Solver, openturns.OptimizationAlgorithm
+
```
Comment (jschueller) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:30]: an example maybe ?

File: python/src/LeastSquaresEquationsSolver_doc.i.in (hunk 7 above)
Comment (sofianehaddad) [reply python/src/LeastSquaresEquationsSolver_doc.i.in:3]: indeed. I realize that for some `Gaussian Process` classes I forgot to add this warning...

Comment (jschueller) [reply python/src/LeastSquaresEquationsSolver_doc.i.in:3]: I can do it in misc if you want

**Comment by regislebrun** [comment]
Any additional work to do before to have this PR in master?

**Comment by sofianehaddad** [comment]
Not on my side

**Review by regislebrun — APPROVED** [review]

**Comment by sofianehaddad** [comment]
@jschueller anything missing for merging this branch ?

**Comment by jschueller** [comment]
no, just mention the new class in the changelog and I'll merge it

**Comment by sofianehaddad** [comment]
@jschueller done

**Comment by sofianehaddad** [comment]
@jschueller done

**Pull request merged**